### ISSUE TITLE
bug: Build fails while converting to DocBook

### ISSUE THREAD

**Issue by dinko-pehar** [nix-community/home-manager#3680]
### Are you following the right branch?

- [X] My Nixpkgs and Home Manager versions are in sync

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Issue description

I have my [dotfiles](https://github.com/dinko-pehar/dotfiles) managed by `darwin-nix` which uses `home-manager`. I've updated the nix channels using `nix-channel --update` but it broke some of the things. I've disabled some of the packages that maybe require some new tweaks to work, but it still fails.

I'm not sure if this is related to Nix or Home-Manager. It seems like it's trying to build some `manual-combined`/ `docbook` for fish completions, but it fails:

```
building '/nix/store/n2902q6psifm4a8yxa683pslsjjczv4r-ytmdl-2022.03.16-fish-completions.drv'...
building '/nix/store/z01252wp181mhql8qlm5ip383afli4xd-zplug-2.4.2-fish-completions.drv'...
building '/nix/store/sh0pf5jkgrxph0fsp91dmk9ww80mki1k-zulu17.34.19-ca-jdk-17.0.3-fish-completions.drv'...
building '/nix/store/vblq2zbhim4yr35sw6556i8jwahlc10q-crystal-1.2.2-fish-completions.drv'...
building '/nix/store/h21xbs4v3qizz1s2y0zmjnmg1kvr2nll-converted-asciidoc.drv'...
building '/nix/store/244lsgm43kbikp37m9phh7gk0h1j1xpd-options.json.drv'...
building '/nix/store/nwpw9fmmn4ffp5yg15j7605i8a8c1bnr-options.json.drv'...
Converting contributing.adoc to DocBook ...
building '/nix/store/v4kia3n7hz99n3780jc880a9bnh1yalg-options-db.xml.drv'...
building '/nix/store/zsvw23xq1sfr0wc68avyq76w0vigvzfw-options-db.xml.drv'...
building '/nix/store/3sg8ra9bbjs5cacqvphk3dxavc9aw2kb-options-db.xml.drv'...
Converting faq.adoc to DocBook ...
Converting installation.adoc to DocBook ...
Converting nix-flakes.adoc to DocBook ...
Converting usage.adoc to DocBook ...
Converting writing-modules.adoc to DocBook ...
Converting release-notes/release-notes.adoc to DocBook ...
Converting release-notes/rl-1809.adoc to DocBook ...
Converting release-notes/rl-1903.adoc to DocBook ...
Converting release-notes/rl-1909.adoc to DocBook ...
Converting release-notes/rl-2003.adoc to DocBook ...
Converting release-notes/rl-2009.adoc to DocBook ...
Converting release-notes/rl-2105.adoc to DocBook ...
Converting release-notes/rl-2111.adoc to DocBook ...
Converting release-notes/rl-2205.adoc to DocBook ...
Converting release-notes/rl-2211.adoc to DocBook ...
Converting release-notes/rl-2305.adoc to DocBook ...
building '/nix/store/n3ap8bmx7bdi8kr1cjlmvfl3wxn9hzzw-nmd-documents.drv'...
created 42 symlinks in user environment
building '/nix/store/zvjbxmv90w44cdq501lc5jnbl2f8sw2w-manual-combined.drv'...

/nix/store/9ff1wzw2fyfj7qmpss1q6sygix4x7m9h-manual-combined/manual-combined.xml:1389: element variablelist: Relax-NG validity error : Did not expect element variablelist there
  1385  </section>
  1386  </chapter>
  1387   <appendix xml:id="ch-options">
  1388    <title>Configuration Options</title>
  1389    <variablelist xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink" xml:id="home-manager-options" xml:base="nmd-result/home-manager-options.xml"><varlistentry><term xlink:href="#opt-_module.args" xml:id="opt-_module.args"><option>_module.args</option></term><listitem><simpara>Additional arguments passed to each module in addition to ones
  1390  like <literal>lib</literal>, <literal>config</literal>,
  1391  and <literal>pkgs</literal>, <literal>modulesPath</literal>.</simpara>

/nix/store/9ff1wzw2fyfj7qmpss1q6sygix4x7m9h-manual-combined/manual-combined.xml:1389: element varlistentry: Relax-NG validity error : Did not expect element varlistentry there
  1385  </section>
  1386  </chapter>
  1387   <appendix xml:id="ch-options">
  1388    <title>Configuration Options</title>
  1389    <variablelist xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink" xml:id="home-manager-options" xml:base="nmd-result/home-manager-options.xml"><varlistentry><term xlink:href="#opt-_module.args" xml:id="opt-_module.args"><option>_module.args</option></term><listitem><simpara>Additional arguments passed to each module in addition to ones
  1390  like <literal>lib</literal>, <literal>config</literal>,
  1391  and <literal>pkgs</literal>, <literal>modulesPath</literal>.</simpara>

/nix/store/9ff1wzw2fyfj7qmpss1q6sygix4x7m9h-manual-combined/manual-combined.xml:1389: element varlistentry: Relax-NG validity error : Expecting element example, got varlistentry
  1385  </section>
  1386  </chapter>
  1387   <appendix xml:id="ch-options">
  1388    <title>Configuration Options</title>
  1389    <variablelist xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink" xml:id="home-manager-options" xml:base="nmd-result/home-manager-options.xml"><varlistentry><term xlink:href="#opt-_module.args" xml:id="opt-_module.args"><option>_module.args</option></term><listitem><simpara>Additional arguments passed to each module in addition to ones
  1390  like <literal>lib</literal>, <literal>config</literal>,
  1391  and <literal>pkgs</literal>, <literal>modulesPath</literal>.</simpara>

/nix/store/9ff1wzw2fyfj7qmpss1q6sygix4x7m9h-manual-combined/manual-combined.xml:1389: element varlistentry: Relax-NG validity error : Expecting element bridgehead, got varlistentry
  1385  </section>
  1386  </chapter>
  1387   <appendix xml:id="ch-options">
  1388    <title>Configuration Options</title>
  1389    <variablelist xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink" xml:id="home-manager-options" xml:base="nmd-result/home-manager-options.xml"><varlistentry><term xlink:href="#opt-_module.args" xml:id="opt-_module.args"><option>_module.args</option></term><listitem><simpara>Additional arguments passed to each module in addition to ones
  1390  like <literal>lib</literal>, <literal>config</literal>,
  1391  and <literal>pkgs</literal>, <literal>modulesPath</literal>.</simpara>

/nix/store/9ff1wzw2fyfj7qmpss1q6sygix4x7m9h-manual-combined/manual-combined.xml:1402: element listitem: Relax-NG validity error : Did not expect element listitem there
  1398  not bound to an attribute.</simpara>
  1399  <simpara>Some arguments are already passed by default, of which the
  1400  following <emphasis>cannot</emphasis> be changed with this option:</simpara>
  1401  <itemizedlist>
  1402  <listitem><para><simpara><varname>lib</varname>: The nixpkgs library.</simpara>
  1403  </para></listitem>
  1404  <listitem><para><simpara><varname>config</varname>: The results of all options after merging the values from all modules together.</simpara>

/nix/store/9ff1wzw2fyfj7qmpss1q6sygix4x7m9h-manual-combined/manual-combined.xml fails to validate
error: builder for '/nix/store/zvjbxmv90w44cdq501lc5jnbl2f8sw2w-manual-combined.drv' failed with exit code 3;
       last 10 log lines:
       > /nix/store/9ff1wzw2fyfj7qmpss1q6sygix4x7m9h-manual-combined/manual-combined.xml:1402: element listitem: Relax-NG validity error : Did not expect element listitem there
       >   1398        not bound to an attribute.</simpara>
       >   1399    <simpara>Some arguments are already passed by default, of which the
       >   1400     following <emphasis>cannot</emphasis> be changed with this option:</simpara>
       >   1401    <itemizedlist>
       >   1402  <listitem><para><simpara><varname>lib</varname>: The nixpkgs library.</simpara>
       >   1403 </para></listitem>
       >   1404      <listitem><para><simpara><varname>config</varname>: The results of all options after merging the values from all modules together.</simpara>
       >
       > /nix/store/9ff1wzw2fyfj7qmpss1q6sygix4x7m9h-manual-combined/manual-combined.xml fails to validate
       For full logs, run 'nix log /nix/store/zvjbxmv90w44cdq501lc5jnbl2f8sw2w-manual-combined.drv'.

error: 1 dependencies of derivation '/nix/store/5j6wkrkz23j5609ydgs613inphcg6y6b-man-pages.drv' failed to build
error: 1 dependencies of derivation '/nix/store/d2qzx8i8pgpm7q3hdp8kycpmq13hqhwa-home-manager-applications.drv' failed to build
error: 1 dependencies of derivation '/nix/store/v93lkfq398x4nd7gry4c1a28bah4kchq-home-manager-fonts.drv' failed to build
error: 1 dependencies of derivation '/nix/store/0lx1cvf22qjpcws74irx9yqcll0084hl-home-manager-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/2ba5a6qhvrin483j8v97bigczrwrrd0b-man-pages-fish-completions.drv' failed to build
error: 1 dependencies of derivation '/nix/store/87nfwaqcgl14if7fl53i2jyg5b6hwvlm-home-manager-generation.drv' failed to build
error: 1 dependencies of derivation '/nix/store/5j7ichyb6gf1a8b5shq90vic0l3ild59-activation-raziel.drv' failed to build
error: 1 dependencies of derivation '/nix/store/nawdrn3dfs0ajcww8aiqqpc5vw4plqyr-darwin-system-23.05pre439589.298add347c2+darwin4.0000000.drv' failed to build
```

### Maintainer CC

Not sure ?

### System information

```markdown
- system: `"aarch64-darwin"`
 - host os: `Darwin 22.2.0, macOS 13.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.12.0`
 - channels(raziel): `"darwin, home-manager, unstable"`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

**Comment by ncfavier** [comment]
This *looks* like a duplicate of https://github.com/NixOS/nixpkgs/issues/196651 (it's hard to tell because DocBook errors are not that informative...)

What branch of home-manager are you following? Are you sure you're on the latest revision?

The following information would be useful:
- home-manager revision
- nixpkgs revision
- full contents of the `/nix/store/9ff1wzw2fyfj7qmpss1q6sygix4x7m9h-manual-combined/manual-combined.xml` file that fails to validate

**Comment by dinko-pehar** [comment]
@ncfavier  Thank you for quick response.

I will update this issue soon when I'm back in few hours with info you requested and try that solution in duplicate you mentioned.

I just updated nix channels so I assume I'm on latest revisions of Home manager, but will update to be sure.

**Comment by dinko-pehar** [comment]
@ncfavier Thank you for your help. Adding the mentioned (`manual.manpages.enable = false;`) makes it work again 🚀 

You asked about revisions, I'm not expert on Nix still but I guess:
- `nix-instantiate --eval -E '(import <nixpkgs> {}).lib.version'`-> `'23.05pre439589.298add347c2'`
- Not sure about `home-manager` and how to evaluate it, but I've run today `nix-channel --update` so I assume it's from todays commit. I can add, for example, Lua config in Neovim. That feature was added few weeks ago.

In conclusion, I guess it is something related to the issue you mentioned so you can close this one because it resolved problems for me.

**Comment by ncfavier** [comment]
Okay, so maybe it is a different issue. Could you paste the contents of that `/nix/store/9ff1wzw2fyfj7qmpss1q6sygix4x7m9h-manual-combined/manual-combined.xml` file to https://gist.github.com/ ?

**Comment by dinko-pehar** [comment]
Here is the content of file you request for further investigation - https://gist.github.com/dinko-pehar/15bccbb2f04dc05746f4ab6375d3e6e0

**Comment by ncfavier** [comment]
Thanks, that's helpful.

That file fails to validate against the DocBook stylesheet because it contains `<simpara>` tags inside `<para>` tags, which isn't allowed (of course, the error couldn't just say that, that would be too simple...).

These errors are only present in the description for `_module.args`:

<details>
<summary> _module.args docbook </summary>

```xml
<simpara>Additional arguments passed to each module in addition to ones
like <literal>lib</literal>, <literal>config</literal>,
and <literal>pkgs</literal>, <literal>modulesPath</literal>.</simpara>
<simpara>This option is also available to all submodules. Submodules do not
inherit args from their parent module, nor do they provide args to
their parent module or sibling submodules. The sole exception to
this is the argument <literal>name</literal> which is provided by
parent modules to a submodule and contains the attribute name
the submodule is bound to, or a unique generated name if it is
not bound to an attribute.</simpara>
<simpara>Some arguments are already passed by default, of which the
following <emphasis>cannot</emphasis> be changed with this option:</simpara>
<itemizedlist>
<listitem><para><simpara><varname>lib</varname>: The nixpkgs library.</simpara>
</para></listitem>
<listitem><para><simpara><varname>config</varname>: The results of all options after merging the values from all modules together.</simpara>
</para></listitem>
<listitem><para><simpara><varname>options</varname>: The options declared in all modules.</simpara>
</para></listitem>
<listitem><para><simpara><varname>specialArgs</varname>: The <literal>specialArgs</literal> argument passed to <literal>evalModules</literal>.</simpara>
</para></listitem>
<listitem><para><simpara>All attributes of <varname>specialArgs</varname></simpara>
<simpara>Whereas option values can generally depend on other option values
thanks to laziness, this does not apply to <literal>imports</literal>, which
must be computed statically before anything else.</simpara>
<simpara>For this reason, callers of the module system can provide <literal>specialArgs</literal>
which are available during import resolution.</simpara>
<simpara>For NixOS, <literal>specialArgs</literal> includes
<varname>modulesPath</varname>, which allows you to import
extra modules from the nixpkgs package tree without having to
somehow make the module aware of the location of the
<literal>nixpkgs</literal> or NixOS directories.</simpara>
<programlisting>
{ modulesPath, ... }: {
  imports = [
    (modulesPath + "/profiles/minimal.nix")
  ];
}
</programlisting></para></listitem>

</itemizedlist><simpara>For NixOS, the default value for this option includes at least this argument:</simpara>
<itemizedlist>
<listitem><para><varname>pkgs</varname>: The nixpkgs package set according to
the <option>nixpkgs.pkgs</option> option.</para></listitem>

</itemizedlist>
<para><emphasis>Type:</emphasis> lazy attribute set of raw value</para><para><emphasis>Declared by:</emphasis></para><simplelist><member><filename xlink:href="https://github.com/nix-community/home-manager/blob/master/lib/modules.nix#blob-path">
                    &lt;home-manager/lib/modules.nix&gt;
                </filename></member></simplelist>
```

</details>

I'm guessing the fix would be to remove the `<para>` tags [here](https://gitlab.com/rycee/nmd/-/blob/master/lib/docbookify-options-json.py#L117).

It is weird that your manual is generating documentation for `_module.args`, though, because that option was made invisible in https://github.com/NixOS/nixpkgs/pull/208407, which is on unstable and 22.11. Could you try updating nixpkgs?

cc @pennae for good measure

**Comment by ncfavier** [comment]
> I'm guessing the fix would be to remove the `<para>` tags [here](https://gitlab.com/rycee/nmd/-/blob/master/lib/docbookify-options-json.py#L117).

Nope, that's not it. Still struggling to find what the error is exactly...

**Comment by spinkyspider** [comment]
So, I'm pretty sure I've got a different bug, but it is manifesting in the same/similar way as this issue. My rebuild switch starts failing at manual-combined.drv and goes on to complain about opt-targets.darwin.defaults."com.apple.Safari".blah blah blah.

But, I'm not running darwin. :-(  I am, however, running on Apple hardware.(Macbook Pro 7,1)

So, I don't know, maybe same, maybe not.  I'm not a coder.  Can someone look over my commands and outputs and let me know if this is
 a new issue or not?

Thanks.



[home-manager.failure-to-install.vim.txt](https://github.com/nix-community/home-manager/files/10810755/home-manager.failure-to-install.vim.txt)

**Comment by dinko-pehar** [comment]
I've just updated all channels and it's still happening. I have disabled it as suggested above by @ncfavier . It's not important to me. If there is any way I could help more with this, please write.

**Comment by rycee** [comment]
@dinko-pehar Could you provide the precise commit hash of the Nixpkgs you use? Something like

    nix-instantiate --eval -E '(import <nixpkgs> {}).lib.version'

may include a hash excerpt, for example.

**Comment by rycee** [comment]
Yeah, the validation error shows up when running

     nix build -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/298add347c2.tar.gz -f. docs.html

but it seems to work with the latest Nixpkgs unstable

    nix build -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/refs/heads/nixpkgs-unstable.tar.gz -f. docs.html

**Comment by ncfavier** [comment]
@spinkyspider that looks like an instance of https://github.com/NixOS/nixpkgs/issues/196651, i.e. caused by the changes in `showOption` in nixpkgs. The fixes for that have landed in both home-manager and nix-darwin a while ago, so I suggest you update everything.

**Comment by ncfavier** [comment]
@dinko-pehar are you sure you're correctly updating the `nixpkgs` channel that nix-darwin uses? Are you running `nix-channel` as root or as your own user?

**Comment by dinko-pehar** [comment]
@ncfavier I've updated as root but now I have problems with `chromium` enabled on darwin. I'll get back to you once I resolve this.

**Comment by dinko-pehar** [comment]
Finally, it's resolved. Many thanks to you for supporting me.

I usually ran `nix-channel --update` and I assumed it updates all channels because it pulled Neovim update from few weeks ago. I didn't do `sudo nix-channel --update` so this was the problem for me I think.
There are too many moving parts while maintaining everything with nix that I need to learn to be able to understand errors better. I'll try in future to help `home-manager` and contribute if I can.

Thank you 👍

**Comment by ncfavier** [comment]
@fricklerhandwerk you asked about Nix UX pain points, I think this is a pretty common one: confusion about root vs. user channels. (Of course, the real problem is that the sane alternative to channels, flakes, are not stable yet...)

**Comment by spinkyspider** [comment]
> @spinkyspider that looks like an instance of [NixOS/nixpkgs#196651](https://github.com/NixOS/nixpkgs/issues/196651), i.e. caused by the changes in `showOption` in nixpkgs. The fixes for that have landed in both home-manager and nix-darwin a while ago, so I suggest you update everything.

@ncfavier Could you clarify, for a nixos newb, what you mean by "update everything"?  In my attached file I thought it shows that I updated everything.  Do you mean I need to update to the bleeding edge Master branch?  Thanks.

**Comment by ncfavier** [comment]
You're tracking the release-21.11 branch of home-manager, which is outdated. Try release-22.11.

The linked issue affects both nix-darwin and home-manager, regardless of platform.

**Comment by seh** [comment]
Since I'm running into this same problem with a fresh installation of Nix on macOS, I tried running both _nix-channel --update_ and _sudo nix-channel --update_. The former succeeds, but the latter warns:
> warning: $HOME ('/Users/seh') is not owned by you, falling back to the one defined in the 'passwd' file ('/var/root')
unpacking channels...
> warning: $HOME ('/Users/seh') is not owned by you, falling back to the one defined in the 'passwd' file ('/var/root')

I can see that in my _~/.nix-defexpr_ directory, there are separate symbolic links: _channels_ and _channels_root_. I'm not sure whether that's pertinent here, though.

I confirmed that my _flake.lock_ file includes the Git commit hashes for the head commits in both the _home-manager_ and _nix-darwin_ repositories. What else might I need to update to get past this DocBook schema conformance problem?

**Comment by ncfavier** [comment]
Why are you using both flakes and channels?

**Comment by seh** [comment]
I’m new enough to Nix that my best answer is, “I don’t know.”

I ran into the problem described here, and have been reading the advice and solutions that worked for others who asked about it before me. That led me to try the _nix-channel_ commands.

**Comment by ncfavier** [comment]
If you're using flakes, as documented [here](https://nix-community.github.io/home-manager/index.html#ch-nix-flakes), then channels should have no incidence at all.

How are you specifying your flake inputs, and what command are you running to get what error?

**Comment by ncfavier** [comment]
Also reported in https://github.com/nix-community/home-manager/issues/4142, this looks like it could be a new instance of this problem. Let's track the new issue there.

**Comment by seh** [comment]
> How are you specifying your flake inputs, and what command are you running to get what error?

Just to answer this question, I'm trying to adapt @sebastiant's [_dotfiles_ repository,](https://github.com/sebastiant/dotfiles) first just targeting one machine running macOS. I'm rewriting as I go, not installing all the programs yet; I'm just trying to get the skeleton working.

**Comment by sebastiant** [comment]
Indeed, channels are not relevant if using flakes. To not hijack the thread: feel free to reach out to me separately if you want assistance with nix or flakes in general.

The corresponding action to update your channel would be `nix flake update` https://nixos.org/manual/nix/unstable/command-ref/new-cli/nix3-flake-update.html which will update all your flake inputs in the flake.lock file. Ill try to reproduce this on my mac from that repository tomorrow!

**Comment by seh** [comment]
> feel free to reach out to me separately if you want assistance with nix or flakes in general.

Where would you prefer to have these discussions that probably don't warrant issues to be filed here? As an example, I now find that the _jsonnet_ package (still using the C++ rather than the Go port) won't build on macOS, due to _ld_ not being able to find symbols for the "x86_64" architecture.
